plugins/modules/win_acl_inheritance.ps1

```diff
@@ -1,75 +1,158 @@
 #!powershell
 
+# Copyright: (c) 2022, Oleg Galushko (@inorangestylee)
 # Copyright: (c) 2015, Hans-Joachim Kliemeck <git@kliemeck.de>
 # GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
 
-#Requires -Module Ansible.ModuleUtils.Legacy
+#AnsibleRequires -CSharpUtil Ansible.Basic
 
-$params = Parse-Args $args -supports_check_mode $true
-$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -default $false
+$spec = @{
+    options = @{
+        path = @{ type = 'str'; required = $true }
+        reorganize = @{ type = 'bool'; default = $false }
+        state = @{ type = 'str'; default = 'absent'; choices = @('absent', 'present') }
+    }
+    supports_check_mode = $true
+}
+
+function Get-AclEx {
+    [CmdletBinding()]
+    [OutputType([System.Security.AccessControl.NativeObjectSecurity])]
+    param(
+        [System.String] $LiteralPath
+    )
 
-$result = @{
-    changed = $false
+    $item = Get-Item -LiteralPath $LiteralPath
+    return $item.GetAccessControl()
 }
 
-$path = Get-AnsibleParam -obj $params "path" -type "path" -failifempty $true
-$state = Get-AnsibleParam -obj $params "state" -type "str" -default "absent" -validateSet "present", "absent" -resultobj $result
-$reorganize = Get-AnsibleParam -obj $params "reorganize" -type "bool" -default $false -resultobj $result
+function Set-AclEx {
+    [CmdletBinding(SupportsShouldProcess)]
+    param(
+        [System.String] $LiteralPath,
+        [System.Security.AccessControl.NativeObjectSecurity] $AclObject
+    )
 
-If (-Not (Test-Path -LiteralPath $path)) {
-    Fail-Json $result "$path file or directory does not exist on the host"
+    $item = Get-Item -LiteralPath $LiteralPath
+    if ($PSCmdlet.ShouldProcess($LiteralPath)) {
+        if ($item.PSProvider.Name -eq 'Registry') {
+            Set-Acl -LiteralPath $LiteralPath -AclObject $AclObject
+        }
+        else {
+            $item.SetAccessControl($AclObject)
```

Comment:
I'm curious why `SetAccessControl` can't be used on a registry key? It seems like the type has the method defined https://docs.microsoft.com/en-us/dotnet/api/microsoft.win32.registrykey.setaccesscontrol?view=net-6.0. Does `Set-Acl` do more work here?

Reply:
I don't know if this is a bug or a feature, but the problem has long been mentioned on various boards. Trying to apply ACL leads to an exception. Most likely the problem is a lack of `SE*` privileges (my guess, as I could not get below the identifier of `Microsoft.Win32.RegistryKey public RegistryKey? OpenSubKey(string name, RegistryRights rights`). `Get-Item` opens the key without `WriteAccess` rights. I've done some research with reflexion and can probably confirm this.

PS7: no WriteAccess flag
```Powershell
$path = "HKLM:\SOFTWARE\Test"
$key = get-item -LiteralPath $path
$state = $key.GetType().GetField("_state", [System.Reflection.BindingFlags]::Instance -bor [System.Reflection.BindingFlags]::NonPublic)
$state.GetValue($key)

0
```

PS5: no WriteAcces flag
```Powershell
$path =  "HKLM:\SOFTWARE\Test"
$key = get-item -LiteralPath $path
$state = $key.GetType().GetField("state", [System.Reflection.BindingFlags]::Instance -bor [System.Reflection.BindingFlags]::NonPublic)
$state.GetValue($key)

0
```

.Net native
```Powershell
# w/o write access flag (default)
$key = [Microsoft.Win32.Registry]::LocalMachine.OpenSubKey('SOFTWARE\Test')
$state = $key.GetType().GetField("_state", [System.Reflection.BindingFlags]::Instance -bor [System.Reflection.BindingFlags]::NonPublic)
$state.GetValue($key)

0

# w/ write access flag
$key = [Microsoft.Win32.Registry]::LocalMachine.OpenSubKey('SOFTWARE\Test', $true)
$state = $key.GetType().GetField("_state", [System.Reflection.BindingFlags]::Instance -bor [System.Reflection.BindingFlags]::NonPublic)
$state.GetValue($key)

WriteAccess
```

Reply:
Ah ok that makes sense that `Get-Item` won't open it with the access required. Just an FYI we do have some shared utils https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/powershell/Ansible.ModuleUtils.PrivilegeUtil.psm1 that can be used to enable privileges on a process as needed.